src/microbit/model/display.py

```diff
@@ -20,33 +23,47 @@ def scroll(self, value, delay=150, wait=True, loop=False, monospace=False):
             )
             thread.start()
             return
-        while True:
-            if isinstance(value, (str, int, float)):
-                value = str(value)
+
+        # Set current_pid to the thread's identifier
+        self.__current_pid = threading.get_ident()
+
+        if isinstance(value, (str, int, float)):
+            value = str(value)
+        else:
+            raise TypeError(f"can't convert {type(value)} object to str implicitly")
+
+        letters = []
+        for c in value:
+            if monospace:
+                letters.append(Display.__get_image_from_char(c))
+                letters.append(
+                    Image(CONSTANTS.SPACE_BETWEEN_LETTERS_WIDTH, CONSTANTS.LED_HEIGHT)
+                )
             else:
-                raise TypeError(f"can't convert {type(value)} object to str implicitly")
-            letters = []
-            for c in value:
-                if monospace:
-                    if c == " ":
-                        letters.append(Image(6, CONSTANTS.LED_HEIGHT))
-                    else:
-                        letters.append(Display.__get_image_from_char(c))
-                        letters.append(Image(1, CONSTANTS.LED_HEIGHT))
+                if c == " ":
+                    letters.append(
+                        Image(CONSTANTS.WHITESPACE_WIDTH, CONSTANTS.LED_HEIGHT)
+                    )
                 else:
-                    if c == " ":
-                        letters.append(Image(3, CONSTANTS.LED_HEIGHT))
-                    else:
-                        letters.append(
-                            Display.__strip_image(Display.__get_image_from_char(c))
+                    letters.append(
+                        Display.__strip_unlit_columns(Display.__get_image_from_char(c))
+                    )
+                    letters.append(
+                        Image(
+                            CONSTANTS.SPACE_BETWEEN_LETTERS_WIDTH, CONSTANTS.LED_HEIGHT,
                         )
-                        letters.append(Image(1, CONSTANTS.LED_HEIGHT))
-            appended_image = Display.__create_scroll_image(letters)
+                    )
+        appended_image = Display.__create_scroll_image(letters)
+
+        while True:
             # Show the scrolled image one square at a time.
             for x in range(appended_image.width() - CONSTANTS.LED_WIDTH + 1):
                 self.__image.blit(
                     appended_image, x, 0, CONSTANTS.LED_WIDTH, CONSTANTS.LED_HEIGHT
                 )
+                # If show or scroll is called again, there will be a different pid and break
+                if self.__current_pid != threading.get_ident():
+                    break
                 time.sleep(delay / 1000)
             if not loop:
                 break

```

Comment:
I don't think the current implementation is thread-safe.
We should fix that :)

Reply:
added locks 🥇